### PR TITLE
API calls for channel opening along with json schema validation

### PR DESCRIPTION
**Pull request by ryle-goehausen**
Not fully complete in validation but enough to make demonstration.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -65,7 +66,10 @@ lazy val coreDependencies = Seq(
   "com.typesafe.akka" %% "akka-remote" % versions.akka,
   "ch.megard" %% "akka-http-cors" % versions.akkaHttpCors,
   "de.heikoseeberger" %% "akka-http-json4s" % "1.16.1",
-  "org.json4s" %% "json4s-native" % "3.6.3",
+  "org.json4s" %% "json4s-native" % "3.6.2",
+  "org.json4s" %% "json4s-ext" % "3.6.2",
+  "org.json4s" %% "json4s-jackson" % "3.6.2",
```
Comment (tylerprete) [inline build.sbt]: this seems weird -- why do we need jackson and native?

File: src/main/scala/org/constellation/API.scala [hunk 2]
```diff
@@ -226,6 +227,19 @@ class API(udpAddress: InetSocketAddress)(implicit system: ActorSystem, val timeo
 
   private val postEndpoints =
     post {
+      pathPrefix("channel") {
+        path("open") {
+          entity(as[ChannelOpenRequest]) { request =>
+            ChannelMessage.createGenesis(request)
+            complete(StatusCodes.OK)
```
Comment (tylerprete) [inline src/main/scala/org/constellation/API.scala]: StatusCode 201 (Created) might be better, though doesn't really matter.

File: src/main/scala/org/constellation/primitives/ChannelMessage.scala [hunk 3]
```diff
@@ -1,15 +1,27 @@
...
+import org.json4s.jackson.JsonMethods.{asJsonNode, parse}
 
+// Should channelId be associated with a unique keyPair or not?
 case class ChannelMessageData(
                                message: String,
                                previousMessageDataHash: String,
                                channelId: String
                              ) extends ProductHash
 
+case class ChannelOpen(
```
Comment (tylerprete) [inline src/main/scala/org/constellation/primitives/ChannelMessage.scala]: arguably better to have some kind of channel trait -- with children like "ValidatedChannel", "BasicChannel" or something. easier to extend to other stuff later as well.

Just an idea though.

File: src/main/scala/org/constellation/primitives/ChannelMessage.scala [hunk 4]
```diff
@@ -30,6 +42,34 @@ object ChannelMessage {
...
+      !dao.threadSafeMessageMemPool.activeChannels.contains(channelOpenRequest.channelId)
+    ) Some {
+
+      val genesisMessageStr = ChannelOpen(channelOpenRequest.jsonSchema, channelOpenRequest.acceptInvalid).json
+      val msg = create(genesisMessageStr, Genesis.CoinBaseHash, channelOpenRequest.channelId)
+      val semaphore = new Semaphore(1)
+      dao.threadSafeMessageMemPool.activeChannels(channelOpenRequest.channelId) = semaphore
+      semaphore.acquire()
+      dao.threadSafeMessageMemPool.put(Seq(msg), overrideLimit = true)
+      msg
```
Comment (tylerprete) [inline src/main/scala/org/constellation/primitives/ChannelMessage.scala]: to me it's a lot more readable when you just do Some(msg) at the end then nest the whole block like this.

File: src/main/scala/org/constellation/primitives/PeerManager.scala [hunk 5]
```diff
@@ -145,7 +145,7 @@ object PeerManager {
   }
 
   def validWithLoopbackGuard(host: String)(implicit dao: DAO): Boolean =
-    (host != dao.externalHostString && host != "127.0.0.1") || !dao.preventLocalhostAsPeer
+    (host != dao.externalHostString && host != "127.0.0.1" && host != "localhost") || !dao.preventLocalhostAsPeer
```
Comment (tylerprete) [inline src/main/scala/org/constellation/primitives/PeerManager.scala]: cool, glad we're getting this in.

File: src/test/scala/org/constellation/cluster/E2ETest.scala [hunk 6]
```diff
@@ -26,7 +27,7 @@ class E2ETest extends AsyncFlatSpecLike with Matchers with BeforeAndAfterAll wit
   override def beforeAll(): Unit = {
     // Cleanup DBs
     //Try{File(tmpDir).delete()}
-    //Try{new java.io.File(tmpDir).mkdirs()}
+    Try{new java.io.File(tmpDir).mkdirs()}
```
Comment (tylerprete) [inline src/test/scala/org/constellation/cluster/E2ETest.scala]: can you use better files File here? Trying to stay consistent with it.

File: src/test/scala/org/constellation/cluster/E2ETest.scala [hunk 7]
```diff
@@ -128,6 +180,51 @@ class E2ETest extends AsyncFlatSpecLike with Matchers with BeforeAndAfterAll wit
...
+        }.headOption
+        (block.soeHash, block.parentSOEHashes, valid, messageParent, messageHash)
+      }
+    }
+
+    // TODO: Duplicate messages appearing sometimes but not others?
+    println(s"Num invalid $numInvalid")
+
+    val ids = messagesInChannelWithBlocks.map{_._1}.zipWithIndex.toMap
+    val msgToBlock = messagesInChannelWithBlocks.flatMap{z => z._5.map{_ -> z._1}}.toMap
```
Comment (tylerprete) [inline src/test/scala/org/constellation/cluster/E2ETest.scala]: I know this is just test code -- but might be worth making a case class for these. 5 members in a tuple starts is a but unwieldy.

File: build.sbt (hunk 1 above)
Comment (ryle-goehausen) [reply build.sbt]: Just came from example for schema validator lib, can change though

File: src/main/scala/org/constellation/API.scala (hunk 2 above)
Comment (ryle-goehausen) [reply src/main/scala/org/constellation/API.scala]: Will fix

File: src/main/scala/org/constellation/primitives/ChannelMessage.scala (hunk 3 above)
Comment (ryle-goehausen) [reply src/main/scala/org/constellation/primitives/ChannelMessage.scala]: I think we'll need to store the channels in a better data structure anyways. I'm actually just putting this as JSON on the message field for the first message which is wrong. We should really extend the edge class better so we can make use of TypedEdgeHash and make distinctions for all the different data types.

File: src/main/scala/org/constellation/primitives/ChannelMessage.scala (hunk 4 above)
Comment (ryle-goehausen) [reply src/main/scala/org/constellation/primitives/ChannelMessage.scala]: Will fix

File: src/main/scala/org/constellation/primitives/EdgeDAO.scala [hunk 8]
```diff
@@ -53,24 +53,26 @@ class ThreadSafeTXMemPool() {
...
 class ThreadSafeMessageMemPool() {
 
-  private var messages = Seq[ChannelMessage]()
+  private var messages = Seq[Seq[ChannelMessage]]()
 
   val activeChannels: TrieMap[String, Semaphore] = TrieMap()
 
+  val messageHashToSendRequest: TrieMap[String, ChannelSendRequest] = TrieMap()
+
   def pull(minCount: Int): Option[Seq[ChannelMessage]] = this.synchronized{
```
[the remark this answers is not in the document]
Comment (ryle-goehausen) [reply src/main/scala/org/constellation/primitives/EdgeDAO.scala]: This was a shortcut for adding batching. I didn't use semaphores on the client batch test only on the randomly generated internal transactions as it assumes the client is polling for completion. That way I can test the external batch submissions faster. There is a reason for this but this needs to be rewritten.

File: src/main/scala/org/constellation/primitives/PeerManager.scala (hunk 5 above)
Comment (ryle-goehausen) [reply src/main/scala/org/constellation/primitives/PeerManager.scala]: Forgot this was on here but yeah.

File: src/test/scala/org/constellation/cluster/E2ETest.scala (hunk 6 above)
Comment (ryle-goehausen) [reply src/test/scala/org/constellation/cluster/E2ETest.scala]: Yeah, theres some other calls I replaced out of paranoia for the file handle open thing that we can change back too now re: snapshots.

File: src/test/scala/org/constellation/cluster/E2ETest.scala (hunk 7 above)
Comment (ryle-goehausen) [reply src/test/scala/org/constellation/cluster/E2ETest.scala]: It needs to be moved to a separate rendering function as well. This doesn't really belong in the test but it's convenient. Can update.

File: build.sbt (hunk 1 above)
Comment (ryle-goehausen) [reply build.sbt]: Oh wait I think this is a jackson specific thing. I'm not sure it can be done with json4s

**Pull request merged**